includes/class-newspack-newsletters-subscription.php

```diff
@@ -402,8 +414,187 @@ public static function add_contact( $contact, $lists = false ) {
 		 */
 		$lists = apply_filters( 'newspack_newsletters_contact_lists', $lists, $contact, $provider->service );
 
-		$errors = new WP_Error();
-		$result = [];
+		if ( true !== $async ) {
+			return self::add_contact_to_provider( $contact, $lists, $is_updating );
+		} else {
+			self::add_subscription_intent( $contact, $lists, $is_updating );
+			$nonce = wp_create_nonce( self::ASYNC_ACTION );
+			$url   = admin_url( 'admin-ajax.php?action=' . self::ASYNC_ACTION . '&nonce=' . $nonce );
+			$args  = [
+				'timeout'   => 0.01,
+				'blocking'  => false,
+				'cookies'   => $_COOKIE, // phpcs:ignore
+				'sslverify' => apply_filters( 'https_local_ssl_verify', false ),
+				'body'      => [
+					'action_name' => self::ASYNC_ACTION,
+					'email'       => $contact['email'],
+				],
+			];
+			wp_remote_post( $url, $args );
+			return true;
+		}
+	}
+
+	/**
+	 * Register a subscription intent.
+	 *
+	 * @param array $contact     Contact information.
+	 * @param array $lists       Array of list IDs to subscribe the contact to.
+	 * @param bool  $is_updating Whether the contact is being updated. If false, the contact is being created.
+	 *
+	 * @return void
+	 */
+	private static function add_subscription_intent( $contact, $lists, $is_updating ) {
+		$intents = get_option( 'newspack_newsletters_subscription_intents', [] );
+		if ( ! is_array( $intents ) ) {
+			$intents = [];
+		}
+		$intents[ wp_hash( $contact['email'] ) ] = [
+			'contact'     => $contact,
+			'lists'       => $lists,
+			'is_updating' => $is_updating,
+			'attempts'    => 0,
+			'errors'      => [],
+		];
+		update_option( 'newspack_newsletters_subscription_intents', $intents );
+	}
+
+	/**
+	 * Get subscription intent.
+	 *
+	 * @param string $email Email address.
+	 *
+	 * @return array|false Subscription intent data or false if not found.
+	 */
+	private static function get_subscription_intent( $email ) {
+		$intents = get_option( 'newspack_newsletters_subscription_intents', [] );
+		if ( empty( $intents ) ) {
+			return $intents;
+		}
+		return $intents[ wp_hash( $email ) ] ?? false;
+	}
+
+	/**
+	 * Remove subscription intent.
+	 *
+	 * @param string $email Email address.
+	 *
+	 * @return void
+	 */
+	private static function remove_subscription_intent( $email ) {
+		$intents = get_option( 'newspack_newsletters_subscription_intents', [] );
+		if ( ! is_array( $intents ) ) {
+			$intents = [];
+		}
+		unset( $intents[ wp_hash( $email ) ] );
+		update_option( 'newspack_newsletters_subscription_intents', $intents );
+	}
+
+	/**
+	 * Process subscription intent.
+	 *
+	 * @param string $email Optional email address to process. It'll process all
+	 *                      intents if email is not provided.
+	 *
+	 * @return void
+	 */
+	public static function process_subscription_intents( $email = null ) {
+		if ( $email ) {
+			$intents = [ self::get_subscription_intent( $email ) ];
+		} else {
+			$intents = get_option( 'newspack_newsletters_subscription_intents', [] );
```

Comment:
I'm always afraid of storing things like this is an option entry.

But not a blocker in this case.. I don't think the subscription cadence is too big

Reply:
There could be an option serving as the intent index and have the intent data spread out in different entries. WDYT?

Reply:
WDYT of simply dispatching another async request and trying again immediately instead of having to store stuff in the DB and doing cron jobs?

If we were to spread the data out we should probably use a CPT... but it feels like too much.

If you want to stick with this implementation, maybe just put a cap of a total amount of intents we will ever store in the option as a fail safe.

Reply:
Storing the intent gives us more room for recovery strategies. If we never store it, there is nothing we can do other than try again immediately which would not cover a short ESP outage.

Not sure about caping the intents. By doing so we are assuming the risk of having a subscription submission going nowhere.

Well... a CPT would add ~10 lines of code to the existing approach. I think it's fine, no?

Reply:
> Storing the intent gives us more room for recovery strategies. If we never store it, there is nothing we can do other than try again immediately which would not cover a short ESP outage.

But if we try three times over the next 3 minutes is not a big difference. Maybe increase the interval then?

> Not sure about caping the intents. By doing so we are assuming the risk of having a subscription submission going nowhere.

If we are confident enough that this will never grow to a point that will OOM the site, we should be confident that this will never grow to a point that a submission will be lost ;) 

> Well... a CPT would add ~10 lines of code to the existing approach. I think it's fine, no?

yeah, I think so